### ISSUE TITLE
Include revert reason details in `eth_estimateGas`

### ISSUE THREAD

**Issue by MicahZoltu** [paritytech/parity-ethereum#11513]
- **Parity Ethereum version**: v2.7.2
- **Operating system**: Docker
- **Installation**: Docker
- **Fully synchronized**: yes
- **Network**: Private Test
- **Restarted**: Yes

An `eth_call` that fails due to a `revert(false, "my reason string")` will return a JSON-RPC error response to the caller and the `error.data` property will contain `Reverted 0x08c379a0...` where the hex value is an ABI encoded method call of `Error(string)` with the reason ("my reason string" in example above) as the string parameter.

An `eth_estimateGas` that fails exactly the same will instead contain an `error.data` property that just contains `Reverted`, but without the encoded error reason string.  In many (almost all?) cases, prior to submitting a transaction an `eth_estimateGas` is executed to find out how much gas the transaction requires.  This also protects the user from submitting a transaction that would just fail once it landed on chain by failing early during the `eth_estimateGas`.  However, without a revert reason, troubleshooting these failures can be incredibly difficult and there is not enough information provided by Parity to give to the user.

**Request**: Make `eth_estimateGas` include an `error.data` property whose contents are identical to that of `eth_call`.

**Comment by dvdplm** [comment]
Definitely makes sense to try to return useful error messages; I think we should do this.

**Comment by niklasad1** [comment]
> Definitely makes sense to try to return useful error messages; I think we should do this.

Yes, I agree but I think the main reason why this is inconsistent is that currently `Client::call` doesn't regard an exception as an `Error` whereas `Client::estimate_gas` does. Thus, `Client::call` does this 
manually and returns more information. You can find the difference [here](https://github.com/OpenEthereum/open-ethereum/blob/master/rpc/src/v1/impls/eth.rs#L990-#L1032)

So ideally, we would align this behavior in `rpc::error` but it is currently not as verbose (or swallows some part of the error not sure).

**Comment by niklasad1** [comment]
I can mentor if anyone is interested in this task

**Comment by imondrag** [comment]
@niklasad1, I'd like to tackle this task as one of my first contributions

**Comment by niklasad1** [comment]
Great, let me know if you need additional help

**Comment by adria0** [comment]
@imondrag I self-assigned the issue to me without being aware that are you working on it, sorry. Do you have any progress on it?

**Comment by adria0** [comment]
It will be nice that #11776 to be implemented as part of this effort, also.

**Comment by k06a** [comment]
I wonder why revert message is missing in trances while it actually is getting returned as returning value.

**Comment by tjayrush** [comment]
> It will be nice that #11776 to be implemented as part of this effort, also.

I was also going to ask if this could be added to the appropriate `trace` routines. We use `trace_block` in our work.
